public/app/features/alerting/unified/hooks/useCombinedRuleNamespaces.ts

```diff
@@ -0,0 +1,130 @@
+import { CombinedRule, CombinedRuleNamespace, Rule, RuleNamespace } from 'app/types/unified-alerting';
+import { RulerRulesConfigDTO } from 'app/types/unified-alerting-dto';
+import { useMemo, useRef } from 'react';
+import { getAllRulesSources, isCloudRulesSource } from '../utils/datasource';
+import { isAlertingRule, isAlertingRulerRule } from '../utils/rules';
+import { useUnifiedAlertingSelector } from './useUnifiedAlertingSelector';
+
+interface CacheValue {
+  promRules?: RuleNamespace[];
+  rulerRules?: RulerRulesConfigDTO | null;
+  result: CombinedRuleNamespace[];
+}
+
+// this little monster combines prometheus rules and ruler rules to produce a unfied data structure
+export function useCombinedRuleNamespaces(): CombinedRuleNamespace[] {
```

Comment:
Wow it is a little monster indeed. But nice work, I'm learning that working with these rules requires a lot of nested callbacks 😓

Reply:
this bit here is called "job security" :laughing:  I spent serious effort to hopefully get it right...